internal/pkg/agent/application/upgrade/artifact/download/http/progress_reporter.go

```diff
@@ -0,0 +1,110 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package http
+
+import (
+	"context"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/atomic"
+)
+
+type downloadProgressReporter struct {
+	sourceURI   string
+	interval    time.Duration
+	warnTimeout time.Duration
+	length      float64
+
+	downloaded atomic.Int
+	started    time.Time
+
+	progressObservers []progressObserver
+}
+
+func newDownloadProgressReporter(sourceURI string, timeout time.Duration, length int, progressObservers ...progressObserver) *downloadProgressReporter {
+	interval := time.Duration(float64(timeout) * downloadProgressIntervalPercentage)
+	if interval == 0 {
+		interval = downloadProgressMinInterval
+	}
+
+	return &downloadProgressReporter{
+		sourceURI:         sourceURI,
+		interval:          interval,
+		warnTimeout:       time.Duration(float64(timeout) * warningProgressIntervalPercentage),
+		length:            float64(length),
+		progressObservers: progressObservers,
+	}
+}
+
+func (dp *downloadProgressReporter) Write(b []byte) (int, error) {
+	n := len(b)
+	dp.downloaded.Add(n)
+	return n, nil
+}
```

Comment:
Why the downloadProgressReporter needs to implement a `Writer` interface that only increments an integer ?

Reply:
This being a refactoring PR, I didn't want to change the implementation too much.  So I didn't touch this.  However, the answer lies in how the progress reporter is informed of a download's progress, over here: https://github.com/elastic/elastic-agent/blob/359be4f18ef1104f60e9199aca268d5dcbfd77f1/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L214